### DISCUSSION TITLE
pattern_reentrant_vi

### DISCUSSION THREAD

**Discussion by giscus** [ruanqizhen/labview_book#13] [Announcements]
# 可重入 VI 和递归算法

https://lv.qizhen.xyz/pattern_reentrant_vi

**Comment by whchrome** [comment]
感谢阮老师，这些内容非常有价值。最近在用LabVIEW编写并行程序，需要调用同一个子vi，才发现子vi的默认设置是非可重入的。函数选版里自带的函数基本都是可重入的，导致我以为自己写的子vi也是可重入的。我个人认为，近些年计算机内存快速增大，价格快速降低，子vi默认非可重入节约的一丢丢内存空间，反倒耽搁了很多运行时间。

**Reply by ruanqizhen** [reply]
非可重入的子 VI，还有一个重要的优点，尤其是对于 LabVIEW 的初学用户来说，它更符合直觉。比如在 [可重入 VI 的副本](https://lv.qizhen.xyz/pattern_reentrant_vi#%E5%8F%AF%E9%87%8D%E5%85%A5-vi-%E7%9A%84%E5%89%AF%E6%9C%AC) 这一段提到的例子，这个  “运行次数” 子 VI 被并行调用了 30 次后，用户本能就会希望最终输出的结果为 30。如果采用可重入 VI，结果就不一定是 30 了。
当然，在完全了解了 LabVIEW 运行机制后，这就不是个问题了。
一个 VI，如果它的输出完全由输入决定，自身没有记录任何可能影响运行结果的状态，那么我也觉得默认为可重入会更好一些。

**Reply by whchrome** [reply]
确实是，各有利弊，感谢回复

**Comment by llj2000** [comment]
利用map缓存递归中的map函数labview中找不到，是我没有这个工具包吗？不过我写了个利用变体作为缓存的，感觉功能一样，确实挺快的

**Reply by ruanqizhen** [reply]
Map 容器是近几年在 LabVIEW 中增加的。我用的 LV 2021，是在函数选板 Programming -> Collection 下面。

**Comment by lixin200111** [comment]
请问阮老师最后那个练习题如何实现，算法思想清楚，但是不知道如何下手

**Reply by ruanqizhen** [reply]
能大致说一下你的算法的步骤吗？

**Reply by lixin200111** [reply]
我的想法是，先固定1为第一个数，然后对剩下的调用本VI递归进行全排列，然后固定2和3为第一个数，对剩下的递归进行全排列。

**Reply by ruanqizhen** [reply]
这个想法很好，按照思路写出程序就可以了。
如果还是写不出，就继续考虑细节：比如要把第 1,2,3 个数分别排在最前，就需要构建一个循环。考虑一下循环次数是多少，每次迭代如何取出数据等。再递归调用，考虑一下需要传递那些数据进入递归函数，何中条件停止递归……

**Comment by Fanhezhao** [comment]
“普通 VI 为什么不能被递归调用”部分第三段中，“但是不能被用于地递归调用。”多个地字

**Comment by ruanqizhen** [comment]
之前博客上的相关文章，搬到这里：

## 如何使用 VI 的重入属性（Reentrant）

   在 VI Properties -> Execution 中可以选择 VI 的Reentrant Execution属性（中文译为：可重入执行）。 我们在《[LabVIEW 程序的内存优化](http://ruanqizhen.spaces.msn.com/blog/cns!5852D4F797C53FB6!1493.entry)》一文中讨论过，尽量不要把 VI 设置为重入属性，因为这样就多占用了内存，降低了运行效率。此外，如果不加注意的话，还可能引发多线程不安全的问题。 尽管可重入 VI 在 LabVIEW 中不是必须的，但是在某些情况下使用可重入 VI 可以简化我们的程序。那么在什么情况下可以使用 Reentrant VI 呢？

    首先看一下图 1 所示的程序，程序中调用的两个子 VI 是同一个 VI，并且不是可重入的 VI。LabVIEW 是自动多线程的语言，那么图中的两个子VI会不会同时执行呢。一定不会的。如果程序中调用的是两个不同的子 VI，LabVIEW 有可能会同时在不同的线程执行它们，但对于两次调用相同的子 VI，LabVIEW 一定要等一个执行完，再执行另一个。

![](http://tkfiles.storage.msn.com/x1pnp_rgmi5o50VQfbvYrXGFYi9Lj2Xxx3Y2HWLvLhJC4ydp8cTXF43mZtd5tntKTCtk0JDEyesVMrkwtmCs4vgoSq9yzofc6WjAyJumKzGw3yhF9cGKFFO3777yPncBH5mYu2T22KgmL4)  
**图1：并行的两个相同子 VI**

    其原因是，LabVIEW 会为每个 VI 都开辟一块内存用于数据存储。作为子 VI，每次被调用，它的局部变量的数据都是被存在同一地址的。与 C 语言相对照，在默认情况下，VI 是不可重入的，VI 中所有的局部变量都是静态变量。如果 LabVIEW 在不同的线程下执行同一 VI，那么两个线程就会同时对这一块数据地址进行读写，就会导致这一块地址内数据的混乱。为避免此类不安全情况的出现，LabVIEW 必须等待一个子 VI 执行结束，再执行另一个子 VI。  
    如果需要图1 中的两个子 VI 同时运行，比如子 VI 所做的工作是读取文件这样一类耗时多、但CPU占用不大的操作，则并行执行可以大大提高效率。这时，就需要把子 VI 设置为可重入了。LabVIEW 在不同的地方调用一个可重入 VI 时，会给它另外分配一个独立的数据地址空间。这样就做到了线程安全。在两个线程执行的子 VI 使用两份在不同的地址存储的数据，也就不会造成混乱。但是千万要注意， 这个“在不同的地方”调用：不可重入的 VI 的局部变量与 C 语言中非静态变量的含义是不同的。在后面提到的计数器的例子可以验证这一点。

     我觉得我说得挺清楚了，出道题目给大家测试一下：

![](http://tkfiles.storage.msn.com/x1pnp_rgmi5o50VQfbvYrXGFYi9Lj2Xxx3YE-i5ACyes_wfwUnfeyJIzqzXCQ-vDcRHxOC5nQiNnrGYdhEOL8RTJGyzhHFkRXdO0Itg7akNljJEJ9g3-BSrH4BWWbyhfDT6i-wbv8Bs_0A)  
**图2：延时子 VI**

![](http://tkfiles.storage.msn.com/x1pnp_rgmi5o50VQfbvYrXGFYi9Lj2Xxx3YSH0i4-GFaCNJc4iW0_x3PjgIOfgOzM3iqPPvngmktMoxGkPRw1PyUqwbtvXOnBylYxvw1Sd2DHvH1NTFCnCgZWTRvsOwPK6w9Np3zEpaJWE)  
**图3：计算延时的主 VI**

    图2 是一个子 VI 的代码，功能是延时 1000 毫秒。图3 是主 VI 的代码，并行调用同一子 VI 两次，并计算程序的执行时间。运行主 VI，total time 的值是多少？  
    答案在文章最后。

    这是可重入 VI 的一种用途，即希望在不同的线程里同时执行同一个子 VI。  
    另外还有一种情况下，也可以用到可重入 VI：即需要使用到子 VI 中局部变量保存的数据，而在不同的调用处，这些数据是独立不同的。这句话可能解释得不那么清楚，看下面例子就会比较容易理解些。

![](http://tkfiles.storage.msn.com/x1pnp_rgmi5o50VQfbvYrXGFYi9Lj2Xxx3YoHJS5_JFy9XozNdvgdPlSc0m9uiJwGFykfKk4B0dZTi2MbxDIK1BSeC72y9q28KfH-YtoPVktb__yLQ1O0qvkC5BESodyyb9-9gcCl8DdZI)  
**图4：计数子 VI**

 ![](http://tkfiles.storage.msn.com/x1pnp_rgmi5o50VQfbvYrXGFYi9Lj2Xxx3Y-6TnMRMsOGjfBzZVD8BdK4koHeAu3r4RW9q4_m_gUV7pz8z-ByclGGTT8CBqS0-rPmnMwKvTE76Sj0wkZRQlpyP6xfaz8ypg)  
**图5：测试计数的主 VI**

    图 4 是一个可重入子 VI 的代码，功能是计算这个VI被运行的次数，每运行一次，输出的 count 值就增加1。图5 是调用它的主VI，用于演示这个计数器。执行主VI一次，output 1 和 output 2 的值分别是 10 和 20，表示这个子 VI 在两处分别被调用了 10 次和 20 次。  
    如果把图 4 中的 VI 改为不可重入，则 output 1 和 output 2 的输出值是不确定的。大家可以自己试一试，再想一下原因。

    当使用递归结构时，参与了递归调用的 VI 是需要被同时调用多次的。因此这些 VI 中的变量必须是局部的，也就是说参与了递归调用的 VI 必须都被设置为可重入。参考：[在 LabVIEW 中实现 VI 的递归调用](http://ruanqizhen.spaces.live.com/blog/cns!5852D4F797C53FB6!1832.entry)

    测试题目答案：如果图2的子 VI 没有设置为可重入，则 total time = 2000；如果设置为可重入则 total time = 1000。

**Reply by ruanqizhen** [reply]
之前博客下的留言：

lcz_1260
2009年03月23日 00:38 编辑
赞一个啦！又学会一种技巧！

回复
Qizhen
2009年02月3日 20:59 编辑
如果你的VI不是很复杂，可以发信给我，我帮你看一下。

回复
颖
2009年01月22日 08:04 编辑
你好，我现在初学Labview,需要调试一个VI,这个Vi曾经能够正常运行的，可是现在总显示有错误，我又找不出来，你能帮我看看吗？

回复
帆
2008年01月17日 10:36 编辑
好强啊 我刚开始学着用~~

回复
Qizhen
2007年07月9日 09:25 编辑
kimi，你可以把你的VI或者截图发给我 ruanqizhen@gmail.com，我看一下才能知道是什么问题。

回复
Zhang Ping
2007年07月7日 15:46 编辑
你好。刚拜读你写的文章，受益匪浅。但是在第二个例子时，图4，计算子VI，我做了一个一样的子VI，但是发现count输出一直是0，我用Probe在各个点进行了测试，程序是运行的，正确的。但是while语句外面的count控件输出一直为零。我特意将while中的i添加了indicator,在while语句内有输出，但在while语句外加上一个indicator，数字仍然没有变化。请问这是什么原因！！！

回复
Unknown
2006年09月16日 19:48 编辑
辛苦了！谢谢

回复
Unknown
2006年08月7日 21:50 编辑
又知道了一样，很高兴，谢谢

回复
清扬
2006年07月22日 16:49 编辑
我也是labview的爱好者啊 以后多多交流呀~

回复
yi
2006年07月21日 21:35 编辑
先收藏，再仔细学。谢谢好文章

回复
onemade
2006年07月21日 01:46 编辑
呵呵　同楼上的

回复
洋
2006年07月20日 20:12 编辑
原来是个LABVIEW高手啊我前段时间的时间就是用LABVIEW写程序HOHO有缘啊以后有这方面的问题就来问你啦~

**Reply by ruanqizhen** [reply]
之前博客下的留言：

Qizhen
2009年08月12日 16:16 编辑
LV2009 可以直接递归，不在类中的VI也可以。

回复
l
2009年08月12日 15:50 编辑
现在LabVIEW2009终于可以递归了,不过是不是必须要使用类才可以呢?

回复
Qizhen
2008年05月23日 18:06 编辑
示例存放的位置已经被更新了

回复
huiyan
2007年09月4日 15:26 编辑
阮前辈,这个怎么下载不了啊
http://ruanqizhen.googlepages.com/Factorial.vi ————递归

回复
斌
2007年02月8日 10:16 编辑
我以为LV不能递归的……，还专门用stack来解决hanoi问题…… 看了你的文章我要再重写一下程序，用两者比较看看！
谢谢！！
